CaviTracer tool for detecting channels, tunnels, pores

karolamik13 opened the pull request:
CaviTracer, created by Eryk Trzcinski, as his MSc work under my supervision.
Additional installation of open3D package is required:
pip install open3d

Test functions:
from prody import *

1. Single PDB:
p = parsePDB('1tqn.pdb')
atoms = p.select("protein")
channels, surface = calcChannels(atoms, output_path='channels_output.pdb')
showCavities(surface)

To create 3D model to display it in ProDy
vmd_path = '/usr/local/bin/vmd'
model = getVmdModel(vmd_path, atoms)

To display channels (all or one by one):
showChannels(channels, surface=surface, model=model)
showChannels(channels, model=model)
getChannelParameters(channels)
showCavities(surface, show_surface=True)

Save all channels independently as PDB:
output = 'chanels_test'
channels, surface = calcChannels(atoms, output, separate=True)

# Display channel 1:
showChannels(channels[1], model)

selected_channels = [channels[1], channels[8]]
showChannels(selected_channels, model)

selected_channels = channels[1:4]
lengths, bottlenecks, volumes = getChannelParameters(selected_channels)
selected_channels_atoms = getChannelAtoms(selected_channels)

atoms_protein = getChannelAtoms(channels, atoms)

Extracting residues that are forming channel[1]:
atoms_protein = getChannelAtoms(channels[1], atoms)
len(atoms_protein.getResnames())

distance = 4
residues = atoms_protein.select('same residue as exwithin '+str(distance)+' of resname FIL')
residues.select('name CA').getResnames(), residues.select('name CA').getResnums()
list(zip(residues.select('name CA').getResnames(), residues.select('name CA').getResnums()))

2. Multiple frames:
p2 = parsePDB('LXmulti.pdb')
channels2, surfaces2 = calcChannelsMultipleFrames(p2)

To display particular channel:
channels2[4][1]

model2 = getVmdModel(vmd_path, p2)
showChannels(channels2[4][1], model2)
showChannels(channels2[4][2], model2)

3. Trajectories:
PDBfile = 'L_traj.pdb'
DCDfile = 'L_traj.dcd'
atoms = parsePDB(PDBfile)
dcd = Trajectory(DCDfile)
dcd.link(atoms)
dcd.setCoords(atoms)

channels3, surfaces3 = calcChannelsMultipleFrames(atoms, dcd)
all_channels, all_surfaces=calcChannelsMultipleFrames(atoms, dcd, output_path = 'channels_dcd', separate=True)

karolamik13 committed:
Merge pull request #6 from prody/main

recent changes in WatFinder from main ProDy

Erykiusz committed:
Work in progress channel detection module

Erykiusz committed:
new commit

Erykiusz committed:
new commit

karolamik13 committed:
Merge branch 'prody:main' into main

Erykiusz committed:
added output for main function and fixed some features

karolamik13 committed:
Merge branch 'prody:main' into main

Erykiusz committed:
Separate visualizer functions

Erykiusz committed:
Fixed calcChannelsMultipleFrames

Erykiusz committed:
added getChannelAtoms

Erykiusz committed:
small fixes

Erykiusz committed:
protein + channels and separate outputs for channels

Erykiusz committed:
fixing documentation

Erykiusz committed:
f-string removed

karolamik13 committed:
Merge pull request #7 from Erykiusz/channels

The first version of a tool for the detection of channels developed by Erykiusz Trzcinski

karolamik13 committed:
Merge branch 'interactions' into prody-cavifinder

karolamik13 committed:
Merge branch 'prody:main' into prody-cavifinder

karolamik13 committed:
Cavi-Finder - fixes for Python 2.7

karolamik13 committed:
Merge branch 'prody-cavifinder' of github.com:karolamik13/ProDy into prody-cavifinder

karolamik13 committed:
Merge branch 'interactions' into prody-cavifinder-updated

karolamik13 committed:
CaviFinder is now separated from InSty

karolamik13 committed:
checkAndImport for Py3 and Py2.7

karolamik13 committed:
getVmdModel compatibility with Python 2.7

karolamik13 committed:
Python 2.7 in calcChannelsMultipleFrames & calcChannels

karolamik13 committed:
PY3K for getChannelAtoms (CaviFinder)

karolamik13 committed:
Merge branch 'prody:main' into prody-cavifinder-reorg

karolamik13 committed:
Channels docs edits

karolamik13 committed:
Additional PY3K in calcChannels

jamesmkrieger committed:
py2 safe stars

karolamik13 committed:
Merge pull request #10 from jamesmkrieger/prody-cavifind-reorg-jk

py2 safe stars

karolamik13 committed:
CaviFinder func in __init__

karolamik13 committed:
CaviFinder - tuple fixes in calcChannels

karolamik13 committed:
selectChannelBySelection()

karolamik13 committed:
selectChannelBySelection, checks added

karolamik13 committed:
selectChannelBySelection - Py2.7 fix

karolamik13 committed:
Merge branch 'prody:main' into prody-cavifinder-reorg

karolamik13 committed:
selectChannelBySelection - residues_file=True, param_file=True options

karolamik13 commented:
New features are added:
Python 3.9.21 | packaged by conda-forge | (main, Dec  5 2024, 13:51:40) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from prody import *
In [2]: atoms = parsePDB('align__addH_7lafA.pdb')
In [3]: selectChannelBySelection(atoms, residue_sele='resid 373', residues_file=True, param_file=True)

selectChannelBySelection() selects channels based on the selection.

I still need to add a function that will return residues per channel into a file (the file is used here) and an additional option to return parameters into a file (also used here).
Those will be available in the next commits.

karolamik13 committed:
getChannelParameters - param_file_name option

karolamik13 committed:
getChannelResidueNames - new function in CaviTracker

karolamik13 committed:
selectChannelBySelection - name unification

karolamik13 committed:
selectChannelBySelection - improvement to fetch longer names correctly

karolamik13 committed:
removed print in CaviTracer

karolamik13 committed:
selectChannelBySelection - final fixes (long names)

karolamik13 committed:
Fix for Beta column - FIL atoms [CaviTracer]

karolamik13 committed:
calcOverlappingSurfaces added

karolamik13 commented:
A new function is now added called calcOverlappingSurfaces()
This function helps to find overlapping parts of the channels across frames or homolog structures.

We can use it:
(1) calcOverlappingSurfaces() - to analyze all PDB files with channels in the current directory
(2) calcOverlappingSurfaces(pdb_files='./test_1') - to analyze only test_1 folder
(3) calcOverlappingSurfaces(pdb_files=['All_channels_for_2sblA.pdb', 'All_channels_for_4wfoA.pdb', 'All_channels_for_2iukB.pdb']) - to analyze the list of selected PDB files with channels generated by CaviTracer.

karolamik13 committed:
calcOverlappingSurfaces() normalization

karolamik13 commented:
calcOverlappingSurfaces() is now normalized [0,1]. It is required when we are analyzing a lot of frames. Otherwise it will not fit to occupancy column.

karolamik13 committed:
calcChannels[pqr format]

karolamik13 committed:
Merge branch 'prody:main' into prody-cavifinder-reorg

karolamik13 committed:
getChannelResidueNames [1letter code option added]

karolamik13 commented:
I updated recent changes in ProDy to be sure that everything is working fine.
Additionally, I added option to save the results as 'pqr' file. 
channels, surface = calcChannels(atoms, output_path='channels_pLoxA_ALL.pqr')
If name is not with '.pdb' or '.pqr', it will be save as 'pdb'.

I added 1letter code in getChannelResidueNames():
getChannelResidueNames(atoms, channels, distA=3, one_letter_aa=True)

['channel0: V86, K87, V89, M374, L378, L383, L425, T442, Q443, A444, A446, G447',
 'channel1: V86, M91, L378, H382, L383, I416, N417, A420, A421, L424, L425, L612',
 'channel2: V86, M91, V189, L193, L378, H382, L383, I416, N417, A420, A421, I423, L424, L425, A428, G429, F430, E604, L612',
 'channel3: P101, W105, E411, G412, F415, I416, N684',
 'channel4: N48, P101, W105, V404, E411, G412, F415, I416, P632, V633, F634, S635, T640, N684',
 'channel5: D47, G62, P101, W105, P398, E411, G412, F415, I416, N684',
 'channel6: F228, A233, F234, Y236, W237, T362, Q365, M565, A569, P570, A571, I572, C573, S576, M595, M596, P597, R669, P670, Y673',
 'channel7: S227, F228, A233, F234, Y236, W237, T362, Q365, A569, P570, A571, I572, C573, S576, M595, M596, P597, R668, R669',
 'channel8: F120, S123, L124, P125, L199, V200, R203, L603',
 'channel9: L116, N119, F120, S123, A214, V215, G217, L219, V606, N607']

karolamik13 committed:
calcChannels - option separate is changed

jamesmkrieger commented:
Otherwise, this all sounds good. Hopefully I can get a chance to try it in a few weeks

karolamik13 commented:
> > I updated recent changes in ProDy to be sure that everything is working fine. Additionally, I added option to save the results as 'pqr' file. channels, surface = calcChannels(atoms, output_path='channels_pLoxA_ALL.pqr') If name is not with '.pdb' or '.pqr', it will be save as 'pdb'.
> 
> I suggest having the default as pqr

Ok. I can do that.

karolamik13 committed:
getChannelParameters() can be applied to single and multiple channel

karolamik13 committed:
calcChannels -> pqr format is default, pdb optional

karolamik13 committed:
Channels -> pqr instead of pdb in all functions

karolamik13 committed:
getChannelParameters() - vizualization of parameters

karolamik13 committed:
print remove

karolamik13 committed:
Changes in names and docs (pqr instead of pdb) [calcOverlappingSurfaces()]

karolamik13 commented:
All the channels will be saved as PQR file. Now we will have:
(1) calcOverlappingSurfaces() - to analyze all PQR files with channels in the current directory
(2) calcOverlappingSurfaces(pqr_files='./test_1') - to analyze only test_1 folder
(3) calcOverlappingSurfaces(pqr_files=['All_channels_for_2sblA.pqr', 'All_channels_for_4wfoA.pqr', 'All_channels_for_2iukB.pqr']) - to analyze the list of selected PQR files with channels generated by CaviTracer.

PQR and PDB files have the same format but when we upload PQR file into VMD it will give us better visualization with radius per predicted bead of the channel. If someone wants PDB just change the name "All_channels_for_2sblA.pqr" to "All_channels_for_2sblA.pdb" and it will work, but then I would recommend to use QuickSurf in VMD and radius 0.5 or 0.4 to see the prediction correctly.

karolamik13 commented:
After changes getChannelParameters() can analyze automatically channels (single PDB) and channels2 (mult-model PDB) or channels3 (trajectory results, see in the previous comments)

karolamik13 commented:
One important thing left: getChannelResidueNames() that will analyze DCD and multi-model PDB. Currently it is working with single PDB file (so you can write a loop and use it anyway for DCD and PDBEnsemble), but I will get there in the upcoming days and provide an automatic function.

karolamik13 committed:
calcOverlappingSurfaces is replaced by calcChannelSurfaceOverlaps()

karolamik13 commented:
Good suggestion James. Thank you. I replaced it.
Now we have:
All the channels will be saved as PQR file. Now we will have:
(1) calcChannelSurfaceOverlaps() - to analyze all PQR files with channels in the current directory
(2) calcChannelSurfaceOverlaps(pqr_files='./test_1') - to analyze only test_1 folder
(3) calcChannelSurfaceOverlaps(pqr_files=['All_channels_for_2sblA.pqr', 'All_channels_for_4wfoA.pqr', 'All_channels_for_2iukB.pqr']) - to analyze the list of selected PQR files with channels generated by CaviTracer.

AnthonyBogetti commented:
@jamesmkrieger did you get a chance to test this?

karolamik13 committed:
Merge branch 'prody:main' into prody-cavifinder-reorg

jamesmkrieger committed:
clean up extra insty line

jamesmkrieger committed:
clean up extra insty line better

jamesmkrieger committed:
add back upper insty line

jamesmkrieger reviewed (changes requested):
Someone still needs to go through and check this in detail, which I haven't managed to do yet

At any rate, it would be good to have unit tests please

karolamik13 commented:
Hi James, 
I plan to do it in this week. I will prepare tutorial in Jupyter Notebook file for other people who would be checking/testing CaviTracer functions.

karolamik13 committed:
Merge branch 'prody:main' into prody-cavifinder-reorg

karolamik13 committed:
Merge branch 'prody:main' into prody-cavifinder-reorg

karolamik13 committed:
Excluding hydrogens from calcChannels()

jamesmkrieger reviewed (commented):
Maybe make that an option

karolamik13 committed:
MOLE 2.0 publication update

karolamik13 committed:
Docs improvements for calcChannels()

karolamik13 committed:
Docs minor change in selectChannelBySelection

karolamik13 committed:
bug fix for PQR format in calcChannelSurfaceOverlaps

karolamik13 committed:
parsePQR import added

karolamik13 committed:
Merge branch 'prody:main' into prody-cavifinder-reorg

karolamik13 committed:
selectChannelBySelection - PQR parser and docs changes

karolamik13 committed:
Changes in docs

karolamik13 committed:
additions for docs calcChannelSurfaceOverlaps

karolamik13 committed:
few changes in docs (cavitracer)

karolamik13 committed:
CaviTracer - getChannelAtoms - fix for multimodel PDB analysis

karolamik13 commented:
The code is still tested and improved when needed, but the first version of the tutorial for the CaviTracer is available here:
https://github.com/prody/ProDy-website/tree/master/tutorials/cavitracer_tutorial
(just click on the selected RST file to display the tutorial)

I am updating it when something changes with the CaviTracer code.

karolamik13 committed:
CaviTracer docs improvements
